src/mlpack/methods/ann/gan/gan_impl.hpp

```diff
@@ -255,111 +225,110 @@ template<
-  discriminator.predictors.cols(numFunctions, numFunctions + batchSize - 1) =
-      boost::apply_visitor(outputParameterVisitor, generator.network.back());
-  discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
-      arma::zeros(1, batchSize);
+  numFunctions = trainData.n_cols;
+  noise.set_size(noiseDim, batchSize);
 
-  // Get the gradients of the Generator.
-  res += discriminator.EvaluateWithGradient(discriminator.parameter,
-      numFunctions, noiseGradientDiscriminator, batchSize);
-  gradientDiscriminator += noiseGradientDiscriminator;
+  // To keep track of where we are.
+  size_t currentFunction = 0;
 
-  if (currentBatch % generatorUpdateStep == 0 && preTrainSize == 0)
-  {
-    // Minimize -log(D(G(noise))).
-    // Pass the error from Discriminator to Generator.
-    discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
-        arma::ones(1, batchSize);
-    discriminator.Gradient(discriminator.parameter, numFunctions,
-        noiseGradientDiscriminator, batchSize);
-    generator.error = boost::apply_visitor(deltaVisitor,
-        discriminator.network[1]);
-
-    generator.Predictors() = noise;
-    generator.ResetGradients(gradientGenerator);
-    generator.Gradient(generator.parameter, 0, gradientGenerator, batchSize);
-
-    gradientGenerator *= multiplier;
-  }
+  // We pass two batches during training hence maxIterations is doubled.
+  discriminatorOptimizer.MaxIterations() =
+      discriminatorOptimizer.MaxIterations() * 2;
 
-  counter++;
-  currentBatch++;
+  // Predictors and responses for generator and discriminator network.
+  arma::mat discriminatorPredictors;
+  arma::mat discriminatorResponses;
 
-  // Revert the counter to zero, if the total dataset get's covered.
-  if (counter * batchSize >= numFunctions)
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; i++)
   {
-    counter = 0;
+    // Is this iteration the start of a sequence?
+    if (currentFunction % numFunctions == 0 && i > 0)
+    {
+      currentFunction = 0;
+    }
+
+    // Find the effective batch size; we have to take the minimum of three
+    // things:
+    // - the batch size can't be larger than the user-specified batch size;
+    // - the batch size can't be larger than the number of functions left.
+    const size_t effectiveBatchSize = std::min(batchSize, numFunctions -
+        currentFunction);
+
+    // Training data for dicriminator.
+    if (effectiveBatchSize != batchSize)
+    {
+      noise.set_size(noiseDim, effectiveBatchSize);
+      discriminatorOptimizer.BatchSize() = effectiveBatchSize;
+      discriminatorOptimizer.MaxIterations() = effectiveBatchSize * 2;
+    }
+    noise.imbue( [&]() { return noiseFunction();} );
+    arma::mat fakeImages;
+    generator.Forward(noise, fakeImages);
+
+    discriminatorPredictors = arma::join_rows(
```

Comment:
Hm, internally the optimiser could call the Shuffle method, which alters the order of the predictors, responses, which I think has to be in sync with discriminatorPredictors. Correct me if I'm wrong. So I think we either have to adjust the Shuffle method. Let me know what you think.

Reply:
Yes, you are quite correct.
In order to fix this, I have now made `discriminatorOptimizer.shuffle()`  and `generator.Optimizer.shuffle()` as `false` before training. This will prevent shuffling.
We will need to add some `documentation`, that the shuffle parameter's in both the optimizer's will be overrided as `false` before `training`.
Let me know if this is not a correct solution.